upstairs/src/downstairs.rs

```diff
@@ -3673,6 +3821,589 @@ impl Downstairs {
+            {
+                Ok(_) => {
+                    info!(log, "notified Nexus of reconcile finish");
+                }
+
+                Err(e) => {
+                    error!(
+                        log,
+                        "failed to notify Nexus of reconcile finish! {e}"
+                    );
+                }
+            }
+        });
+    }
+
+    #[cfg(feature = "omicron-build")]
+    fn notify_nexus_of_reconcile_progress(
+        &mut self,
+        reconcile_id: Uuid,
+        current_task: usize,
+        task_count: usize,
+    ) {
+        use crate::get_nexus_client;
+        use chrono::Utc;
+        use nexus_client::types::RepairProgress;
+        use omicron_uuid_kinds::GenericUuid;
+        use omicron_uuid_kinds::TypedUuid;
+        use omicron_uuid_kinds::UpstairsKind;
+        use omicron_uuid_kinds::UpstairsRepairKind;
+
+        let upstairs_id: TypedUuid<UpstairsKind> =
+            TypedUuid::from_untyped_uuid(self.cfg.upstairs_id);
+        let repair_id: TypedUuid<UpstairsRepairKind> =
+            TypedUuid::from_untyped_uuid(reconcile_id);
+
+        let now = Utc::now();
+        let log = self.log.new(o!("reconcile" => repair_id.to_string()));
+
+        // Spawn a task so we don't block the main loop talking to
+        // Nexus.
+        let target_addrs = self.get_target_addrs();
+        tokio::spawn(async move {
+            let Some(nexus_client) =
+                get_nexus_client(&log, &target_addrs).await
+            else {
+                // Exit if no Nexus client returned from DNS - our notification
+                // is best effort.
+                return;
+            };
+
+            match omicron_common::retry_until_known_result(&log, || async {
+                nexus_client
+                    .cpapi_upstairs_repair_progress(
+                        &upstairs_id,
+                        &repair_id,
+                        &RepairProgress {
+                            time: now,
+                            // surely we won't have usize::MAX extents
+                            current_item: current_task as i64,
+                            // i am serious, and don't call me shirley
+                            total_items: task_count as i64,
+                        },
+                    )
+                    .await
+            })
+            .await
+            {
+                Ok(_) => {
+                    info!(log, "notified Nexus of reconcile progress");
+                }
+
+                Err(e) => {
+                    error!(
+                        log,
+                        "failed to notify Nexus of reconcile progress! {e}"
+                    );
+                }
+            }
+        });
+    }
+
+    #[cfg(feature = "omicron-build")]
+    pub(crate) fn notify_nexus_of_client_task_stopped(
+        &mut self,
+        client_id: ClientId,
+        reason: ClientRunResult,
+    ) {
+        use crate::get_nexus_client;
```

Comment:
You could use `cfg_if!` to do all of these imports in bulk at the top of the module, e.g.

```rust
cfg_if::cfg_if! {
    if #[cfg(feature = "omicron-build")] {
        use chrono::Utc;
        use nexus_client::types::DownstairsClientStopped;
        use nexus_client::types::DownstairsClientStoppedReason;
        use omicron_uuid_kinds::DownstairsKind;
        use omicron_uuid_kinds::GenericUuid;
        use omicron_uuid_kinds::TypedUuid;
        use omicron_uuid_kinds::UpstairsKind;
    }
}
```

It's an extra dependency, but is already in our tree.

Reply:
VERY COOL, done in 2644080, though I didn't find it in our tree already?

Reply:
Ah yeah, I saw it in `Cargo.lock`, not `Cargo.toml`, so it was in the _extended_ dependency tree 😄